akka-http-tests/src/test/scala/akka/http/javadsl/server/HttpAppSpec.scala

```diff
@@ -22,48 +23,55 @@ import scala.runtime.BoxedUnit
 class HttpAppSpec extends AkkaSpec with RequestBuilding with Eventually {
   import system.dispatcher
 
-  "HttpApp" should {
+  "HttpApp  Java" should {
 
     "start without ActorSystem" in {
-      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
 
       val minimal = new MinimalHttpApp()
 
       val server = Future {
-        minimal.startServer(host, port, ServerSettings.create(ConfigFactory.load))
+        minimal.startServer("localhost", 0, ServerSettings.create(ConfigFactory.load))
       }
 
-      callAndVerify(host, port, "foo")
+      val binding = minimal.bindingPromise.get(5, TimeUnit.SECONDS)
 
-      // Requesting the server to shutdown
-      callAndVerify(host, port, "shutdown")
+      try {
+        // Requesting the server to shutdown
+        callAndVerify(binding.localAddress.getHostName, binding.localAddress.getPort, "shutdown")
 
-      Await.ready(server, Duration(1, TimeUnit.SECONDS))
-      server.isCompleted should ===(true)
+        Await.ready(server, Duration(1, TimeUnit.SECONDS))
+        server.isCompleted should ===(true)
+      } finally {
+        if (!minimal.shutdownTrigger.isDone())
+          minimal.shutdownTrigger.complete(Done)
+      }
 
     }
 
     "start providing an ActorSystem" in {
-      val (_, host, port) = TestUtils.temporaryServerHostnameAndPort()
 
       val minimal = new MinimalHttpApp()
 
       val server = Future {
-        minimal.startServer(host, port, ServerSettings.create(system), system)
+        minimal.startServer("localhost", 0, ServerSettings.create(system), system)
       }
 
-      callAndVerify(host, port, "foo")
+      val binding = minimal.bindingPromise.get(5, TimeUnit.SECONDS)
 
-      // Requesting the server to shutdown
-      callAndVerify(host, port, "shutdown")
+      try {
+        // Requesting the server to shutdown
+        callAndVerify(binding.localAddress.getHostName, binding.localAddress.getPort, "shutdown")
 
-      Await.ready(server, Duration(1, TimeUnit.SECONDS))
-      server.isCompleted should ===(true)
-      system.whenTerminated.isCompleted should ===(false)
+        Await.ready(server, Duration(1, TimeUnit.SECONDS))
```

Comment:
Given the amount of additional code to ensure proper shutdown would it make sense to also change tests to either use test fixtures/contexts or before/after hooks to handle this?